### ISSUE TITLE
Cargo generates lockfile with duplicate package

### ISSUE THREAD

**Issue by Zoxc** [rust-lang/cargo#2981]
When rebuilding [this project](https://github.com/AveryOS/avery/blob/master/vendor/cargo-sysroot/Cargo.toml) which overrides rustc's `libc_shim` I get the error `thread 'main' panicked at 'a package was referenced twice in the lockfile', src/cargo\core\resolver\encode.rs:51`

[The lockfile with multiple libc packages](https://gist.githubusercontent.com/Zoxc/298955a7c906ef1877f1c8e7753026a4/raw/7a171d3f65d63716667cf32da176cfc0cea3e55f/gistfile1.txt)

This is on [my cargo branch](https://github.com/AveryOS/cargo) based on e95828103a1d8be1161adf394434b18e2e708954

**Comment by bkchr** [comment]
Hi, I have the same problem. I'm having the following project structure:
Masterproject uses ProjectA and ProjectB. ProjectB and ProjectA both using ProjectC(same version for both). Now cargo generates a Cargo.lock file with ProjectC being present two times. All of these projects are included by being git submodules and the Cargo.toml uses the path to the submodule.

My cargo version: cargo 0.19.0 (28d1d60d4 2017-05-16)

Removing the duplicate entry by hand let's cargo compile the project, but the duplicate is added again..

**Comment by epage** [comment]
Does anyone have a minimal reproduction case for this?

I tried loading up the avery one in case I could reproduce with HEAD but there are other setup steps that I didn't want to guess my way through as then I wouldn't be sure if  see no bug if cargo is fixed or if I did something wrong.

In avery's case, a `replace` is being used.  Unsure about the other case.  I'd be surprised if these are still busted 7 years later and wonder if we should close if we there aren't more details to go off of.
